rust/kernel/device.rs

```diff
@@ -74,3 +84,123 @@ impl Drop for Device {
         unsafe { bindings::put_device(self.ptr) };
     }
 }
+
+/// Device data.
+///
+/// When a device is removed (for whatever reason, for example, because the device was unplugged or
+/// because the user decided to remove the driver), the driver is given a chance to clean its state
+/// up, and all io resources should ideally not be used anymore.
+///
+/// However, the device data is reference-counted because other subsystems hold pointers to it. So
+/// some device state must be freed and not used anymore, while others must remain accessible.
+///
+/// This struct separates the device data into three categories:
+/// 1. Io resources: are available until the device is removed.
+/// 2. Registrations: are destroyed on when the device is removed, but before the io resources
+///    become inaccessible.
+/// 3. General data: remain available as long as the ref count is nonzero.
+///
+/// This struct implements the `DeviceRemoval` trait so that it can clean resources up even if not
+/// explicitly called by the device drivers.
+pub struct Data<Reg, Res, Gen> {
+    regs: RevocableMutex<Reg>,
+    res: Revocable<Res>,
+    general: Gen,
+}
+
+/// Safely creates an new reference-counted instance of [`Data`].
+#[doc(hidden)]
+#[macro_export]
+macro_rules! new_device_data {
+    ($reg:expr, $res:expr, $gen:expr, $name:literal) => {{
+        static mut CLASS1: core::mem::MaybeUninit<$crate::bindings::lock_class_key> =
+            core::mem::MaybeUninit::uninit();
+        static mut CLASS2: core::mem::MaybeUninit<$crate::bindings::lock_class_key> =
+            core::mem::MaybeUninit::uninit();
+        let regs = $reg;
+        let res = $res;
+        let gen = $gen;
+        let name = $crate::c_str!($name);
+        // SAFETY: `CLASS1` and `CLASS2` are never used by Rust code directly; the C portion of the
+        // kernel may change it though.
+        #[allow(unused_unsafe)]
```

Comment:
The unsafe is used here, right?

Reply:
This was because we had unsafe blocks in `res` and `gen`, but I moved them to their own biding above (to ensure callers had to wrap them in `unsafe` blocks in such cases), so this indeed isn't needed anymore. Removed.